server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java

```diff
@@ -141,144 +144,150 @@ static boolean isIdleThread(ThreadInfo threadInfo) {
         }
-        for (int t = 0; t < busiestThreads; t++) {
-            long time = getter.applyAsLong(hotties.get(t));
-            String threadName = null;
-            for (ThreadInfo[] info : allInfos) {
-                if (info != null && info[t] != null) {
-                    if (ignoreIdleThreads && isIdleThread(info[t])) {
-                        info[t] = null;
-                        continue;
+
+        return result;
+    }
+
+    private void setThreadWaitBlockTimeMonitoringEnabled(ThreadMXBean threadBean, boolean enabled) {
+        if (threadBean.isThreadContentionMonitoringSupported()) {
+            AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                threadBean.setThreadContentionMonitoringEnabled(enabled);
+                return null;
+            });
+        }
+    }
+
+    String innerDetect(ThreadMXBean threadBean, long currentThreadId) throws Exception {
+        if (threadBean.isThreadCpuTimeSupported() == false) {
+            throw new ElasticsearchException("thread CPU time is not supported on this JDK");
+        }
+
+        StringBuilder sb = new StringBuilder()
+            .append("Hot threads at ")
+            .append(DATE_TIME_FORMATTER.format(LocalDateTime.now(Clock.systemUTC())))
+            .append(", interval=")
+            .append(interval)
+            .append(", busiestThreads=")
+            .append(busiestThreads)
+            .append(", ignoreIdleThreads=")
+            .append(ignoreIdleThreads)
+            .append(":\n");
+
+        // Enabling thread contention monitoring is required for capturing JVM thread wait/blocked times
+        setThreadWaitBlockTimeMonitoringEnabled(threadBean, true);
+
+        try {
+            // Capture before and after thread state with timings
+            Map<Long, ThreadTimeAccumulator> previousThreadInfos = getAllValidThreadInfos(threadBean, currentThreadId);
+            Thread.sleep(interval.millis());
+            Map<Long, ThreadTimeAccumulator> latestThreadInfos = getAllValidThreadInfos(threadBean, currentThreadId);
+
+            latestThreadInfos.forEach((threadId, accumulator) -> accumulator.subtractPrevious(previousThreadInfos.get(threadId)));
+
+            // Sort by delta CPU time on thread.
+            List<ThreadTimeAccumulator> topThreads = new ArrayList<>(latestThreadInfos.values());
+            final ToLongFunction<ThreadTimeAccumulator> getter = ThreadTimeAccumulator.valueGetterForReportType(type);
+
+            CollectionUtil.introSort(topThreads, Comparator.comparingLong(getter).reversed());
+            topThreads = topThreads.subList(0, Math.min(busiestThreads, topThreads.size()));
+            long[] topThreadIds = topThreads.stream().mapToLong(t -> t.threadId).toArray();
+
+            // analyse N stack traces for the top threads
+            ThreadInfo[][] allInfos = captureThreadStacks(threadBean, topThreadIds);
+
+            for (int t = 0; t < topThreads.size(); t++) {
```

Comment:
The code in this nested 'for' loop mostly remained the same, except for simplification of the loop at line 273.